pkg/controller/serviceaccount/tokengetter.go

```diff
@@ -20,30 +20,43 @@ import (
 	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	clientset "k8s.io/client-go/kubernetes"
+	v1listers "k8s.io/client-go/listers/core/v1"
 	"k8s.io/kubernetes/pkg/serviceaccount"
 )
 
 // clientGetter implements ServiceAccountTokenGetter using a clientset.Interface
 type clientGetter struct {
-	client clientset.Interface
+	client               clientset.Interface
+	secretLister         v1listers.SecretLister
+	serviceAccountLister v1listers.ServiceAccountLister
+	podLister            v1listers.PodLister
 }
 
 // NewGetterFromClient returns a ServiceAccountTokenGetter that
 // uses the specified client to retrieve service accounts and secrets.
 // The client should NOT authenticate using a service account token
 // the returned getter will be used to retrieve, or recursion will result.
-func NewGetterFromClient(c clientset.Interface) serviceaccount.ServiceAccountTokenGetter {
-	return clientGetter{c}
+func NewGetterFromClient(c clientset.Interface, secretLister v1listers.SecretLister, serviceAccountLister v1listers.ServiceAccountLister, podLister v1listers.PodLister) serviceaccount.ServiceAccountTokenGetter {
+	return clientGetter{c, secretLister, serviceAccountLister, podLister}
 }
 
 func (c clientGetter) GetServiceAccount(namespace, name string) (*v1.ServiceAccount, error) {
+	if serviceAccount, err := c.serviceAccountLister.ServiceAccounts(namespace).Get(name); err == nil {
+		return serviceAccount, nil
+	}
 	return c.client.CoreV1().ServiceAccounts(namespace).Get(name, metav1.GetOptions{})
 }
 
 func (c clientGetter) GetPod(namespace, name string) (*v1.Pod, error) {
+	if pod, err := c.podLister.Pods(namespace).Get(name); err == nil {
+		return pod, nil
+	}
 	return c.client.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
 }
 
 func (c clientGetter) GetSecret(namespace, name string) (*v1.Secret, error) {
+	if secret, err := c.secretLister.Secrets(namespace).Get(name); err == nil {
+		return secret, nil
+	}
```

Comment:
isn't the recommendation for informer based controllers to always use listers(which extracts from the informer cache) rather than going directly to the apiserver using clients ? Should we just remove the direct call ?

Reply:
> Should we just remove the direct call?

No, creating a token then immediately making an API call with it should be reliable which means falling back to the lookup